src/main/java/org/opensearch/neuralsearch/query/NeuralKNNQueryBuilder.java

```diff
@@ -0,0 +1,158 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.opensearch.neuralsearch.query;
+
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.QueryVisitor;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.search.Weight;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.xcontent.XContentBuilder;
+import org.opensearch.index.query.AbstractQueryBuilder;
+import org.opensearch.index.query.QueryBuilder;
+import org.opensearch.index.query.QueryRewriteContext;
+import org.opensearch.index.query.QueryShardContext;
+import org.opensearch.knn.index.query.KNNQueryBuilder;
+import org.opensearch.knn.index.query.rescore.RescoreContext;
+
+import java.io.IOException;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * NeuralKNNQueryBuilder wraps KNNQueryBuilder to:
+ * 1. Isolate KNN plugin API changes to a single location
+ * 2. Allow extension with neural-search-specific information (e.g., query text)
+ */
+public class NeuralKNNQueryBuilder extends AbstractQueryBuilder<NeuralKNNQueryBuilder> {
+    private final KNNQueryBuilder knnQueryBuilder;
+
+    /**
+     * Wraps KNN Lucene query to support neural search extensions.
+     * Delegates core operations to the underlying KNN query.
+     */
+    public static class NeuralKNNQuery extends Query {
```

Comment:
Can we have this in a separate file? Is there any reason why we want to have it inside builder?

Reply:
It's already a separate new file in NeuralKNNQueryBuilder

Reply:
I mean, moving this class to `NeuralKNNQuery.java`

Reply:
I see, will update to separate NeuralKNNQuery out to make it cleaner